server/src/test/java/keywhiz/service/permission/OwnershipPermissionCheckTest.java

```diff
@@ -0,0 +1,196 @@
+package keywhiz.service.permission;
+
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.collect.ImmutableSet;
+import java.util.Base64;
+import keywhiz.api.ApiDate;
+import keywhiz.api.model.Client;
+import keywhiz.api.model.Group;
+import keywhiz.api.model.Secret;
+import keywhiz.auth.User;
+import keywhiz.service.daos.AclDAO;
+import keywhiz.service.permissions.Action;
+import keywhiz.service.permissions.OwnershipPermissionCheck;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.Mockito.when;
+
+public class OwnershipPermissionCheckTest {
+  @Rule public MockitoRule mockito = MockitoJUnit.rule();
+
+  @Mock AclDAO.AclDAOFactory aclDAOFactory;
+  @Mock AclDAO aclDAO;
+
+  private MetricRegistry metricRegistry;
+  private OwnershipPermissionCheck ownershipPermissionCheck;
+
+  private static final ApiDate NOW = ApiDate.now();
+  private static final Secret.LazyString ENCRYPTED_SECRET = () -> Base64.getUrlEncoder().encodeToString("content".getBytes(UTF_8));
+
+  private static final Secret SECRET_WITH_OWNER = new Secret(0, "secret.with.owner", "owner", null,
+      ENCRYPTED_SECRET, "checksum", NOW, null, NOW, null,
+      null, null, null, 0, null, null, null);
+  private static final Secret SECRET_WITHOUT_OWNER = new Secret(1, "secret.without.owner", null, null,
+      ENCRYPTED_SECRET, "checksum", NOW, null, NOW, null,
+      null, null, null, 0, null, null, null);
+
+  private static final Client CLIENT = new Client(0, "client", null, null, null,
+      null, null, null, null, null, false, true);
+  private static final User USER = User.named("user");
+
+  private static final Group SECRET_OWNER_GROUP = new Group(0, "owner", null, NOW, null, NOW, null, null);
+  private static final Group NOT_SECRET_OWNER_GROUP = new Group(1, "not-owner", null, NOW, null, NOW, null, null);
+
+  private static final String SUCCESS_METRIC_NAME = MetricRegistry.name(OwnershipPermissionCheck.class, "success", "histogram");
+  private static final String FAILURE_METRIC_NAME = MetricRegistry.name(OwnershipPermissionCheck.class, "failure", "histogram");
+
+  @Before
+  public void setUp() {
+    metricRegistry = new MetricRegistry();
+    when(aclDAOFactory.readwrite()).thenReturn(aclDAO);
+    ownershipPermissionCheck = new OwnershipPermissionCheck(metricRegistry, aclDAOFactory);
+  }
+
+  @Test
+  public void testIsAllowedWhenSourceIsNotAClient() {
+    boolean permitted = ownershipPermissionCheck.isAllowed(USER, Action.CREATE, SECRET_WITH_OWNER);
+
+    assertThat(permitted).isEqualTo(false);
```

Comment:
Had to make this explicitly equal to false because it appears that `assertThat(permitted)` always returns true